### PR TITLE
Added User Info endpoint (#1133)

### PR DESCRIPTION
**Pull request by kalinon**
Resolved merge conflicts for #1133

### CODE REVIEW

**Comment by kalinon** [comment]
> @kalinon How can I help to get this merged?

Not really up to me. Needs reviewers.

File: connector/connector.go [hunk 1]
```diff
@@ -97,3 +97,8 @@ type RefreshConnector interface {
 	// changes since the token was last refreshed.
 	Refresh(ctx context.Context, s Scopes, identity Identity) (Identity, error)
 }
+
+// UserInfoConnector represents connectors that support the user info endpoint
+type UserInfoConnector interface {
+	GetUserInfo(connData []byte, user *map[string]interface{}) (err error)
```
Comment (srenatus) [inline connector/connector.go]: [nit] let's not name the return parameter here. While there's no consistent style in the other interfaces above, if all we return is `error`, there's nothing this helps explaining. 😉 What I mean is that above, [for example in HandlePOST](https://github.com/dexidp/dex/pull/1315/files#diff-dac29093b414e413c0bbc846e11d8101R90), there's _two `string` parameters_, so giving them names is illuminating. For one error, it's not. Likewise for [Refresh](https://github.com/dexidp/dex/pull/1315/files#diff-dac29093b414e413c0bbc846e11d8101R98).)

File: connector/connector.go [hunk 2]
```diff
@@ -97,3 +97,8 @@ type RefreshConnector interface {
 	// changes since the token was last refreshed.
 	Refresh(ctx context.Context, s Scopes, identity Identity) (Identity, error)
 }
+
+// UserInfoConnector represents connectors that support the user info endpoint
```
Comment (srenatus) [inline connector/connector.go]: Hmm that's interesting. We're relaying the connector's userinfo data! I would have assumed we'd have a single userinfo endpoint in Dex, which merely decodes the ID token info, or renders the identity as far as Dex is concerned 🤔

File: server/handlers.go [hunk 3]
```diff
@@ -660,7 +663,28 @@ func (s *Server) handleAuthCode(w http.ResponseWriter, r *http.Request, client s
...
+		Email:         authCode.Claims.Email,
+		EmailVerified: authCode.Claims.EmailVerified,
+		Groups:        authCode.Claims.Groups,
+		ConnectorData: string(authCode.ConnectorData),
+		ConnectorId:   authCode.ConnectorID,
+		Expiry:        authCode.Expiry.Unix(),
+	}
+
+	// The accessToken is build from claims
+	accessToken, err := internal.Marshal(token)
```
Comment (srenatus) [inline server/handlers.go]: :x: unchecked error

File: server/handlers.go [hunk 4]
```diff
@@ -998,6 +1023,84 @@ func (s *Server) writeAccessToken(w http.ResponseWriter, idToken, accessToken, r
...
+	decryptedToken, err := s.Decrypt(token)
+	if err != nil {
+		s.logger.Errorf("Internal server error: %v", err)
+		s.renderError(w, http.StatusInternalServerError, "Internal server error.")
+		return
+	}
+
+	accessToken := new(internal.AccessToken)
+	if err := internal.Unmarshal(decryptedToken, accessToken); err != nil {
+		s.logger.Errorf("Failed to unmarshal access token: %v", err)
```
Comment (srenatus) [inline server/handlers.go]: ❓ log and ignore is good enough for this one?

File: server/handlers.go [hunk 5]
```diff
@@ -998,6 +1023,84 @@ func (s *Server) writeAccessToken(w http.ResponseWriter, idToken, accessToken, r
...
+	// Delegate the responsibility of coming up with the user info
+	// to the connector if it is a UserInfoConnector
+	// e.g. oidc connector
+	case connector.UserInfoConnector:
+
+		connData := []byte(accessToken.ConnectorData)
+		err = conn.GetUserInfo(connData, &user)
+
+		if err != nil {
+			s.logger.Errorf("Failed to retrieve user info: %v", err)
```
Comment (srenatus) [inline server/handlers.go]: This is also log-and-ignore. Let's avoid that -- the call `conn.GetUserInfo(connData, &user)` could have done anything to `user`, like, have it filled half-way.

Let's do something like this
```go
extraInfo, err := conn.GetUserInfo(connData)
if err != nil {
    // log, return 500, etc -- but definitely `return`
}
for k, v := range extraInfo {
    user[k] = v
}
```
What do you think?

File: connector/connector.go (hunk 2 above)
Comment (srenatus) [reply connector/connector.go]: OK, I've looked at the `/userinfo` handler code now -- this allows connectors to _enrich_ the user info response, right?

File: connector/oidc/oidc.go [hunk 6]
```diff
@@ -215,10 +225,57 @@ func (c *oidcConnector) HandleCallback(s connector.Scopes, r *http.Request) (ide
...
+
+	// Prepare http client to execute get request
+	client := &http.Client{}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return fmt.Errorf("Error Executing GET request: %v", err)
+	}
+
+	json.NewDecoder(resp.Body).Decode(user)
```
Comment (srenatus) [inline connector/oidc/oidc.go]: :x: unchecked error

File: server/handlers.go [hunk 7]
```diff
@@ -998,6 +1023,84 @@ func (s *Server) writeAccessToken(w http.ResponseWriter, idToken, accessToken, r
 	w.Write(data)
 }
 
+// Handler for the /userinfo endpoint
+func (s *Server) handleUserInfo(w http.ResponseWriter, r *http.Request) {
+
+	token := ""
+	if r.Header.Get("Authorization") != "" {
```
Comment (srenatus) [inline server/handlers.go]: [nit] let's use some local vars, like
```go
if val := r.Header.Get("Authorization"); val != "" {
    subs := strings.Split(val, " ")
    token = subs[1] // TODO: do we know that it's not out-of-bounds array indexing? i.e. len(subs) >= 2?
} else if val := r.URL.Query().Get("access_token"); val != "" {
    token = val
} else {
    s.logger.Error("No token found in header or query")
    return
}
```

File: server/handlers.go [hunk 8]
```diff
@@ -974,6 +998,7 @@ func (s *Server) writeAccessToken(w http.ResponseWriter, idToken, accessToken, r
 	// TODO(ericchiang): figure out an access token story and support the user info
```
Comment (srenatus) [inline server/handlers.go]: This comment can go away now 😃

File: server/security.go [hunk 9]
```diff
@@ -0,0 +1,74 @@
+package server
```
Comment (srenatus) [inline server/security.go]: ℹ️ haven't reviewed this yet

File: .gitignore [hunk 10]
```diff
@@ -1,3 +1,4 @@
 bin
 dist
 _output
+.idea
```
Comment (JoelSpeed) [inline .gitignore]: Nit: This file should probably retain the new line at the end

File: connector/oidc/oidc.go [hunk 11]
```diff
@@ -215,10 +225,57 @@ func (c *oidcConnector) HandleCallback(s connector.Scopes, r *http.Request) (ide
...
+	if err := json.Unmarshal(connData, &cData); err != nil {
+		c.logger.Errorf("Failed to read connector data : %v", err)
+		return err
+	}
+
+	// Format Authorization header to request user info on behalf of the client
+	authorization := fmt.Sprintf("bearer %s", cData.AccessToken)
+	// Prepare get request including Authorization header from RP
+	req, err := http.NewRequest("GET", c.userInfoURI, nil)
+
```
Comment (JoelSpeed) [inline connector/oidc/oidc.go]: Nit: This blank line is kinda odd, would expect the `if err != nil` to directly follow where the error is created

File: server/handlers.go [hunk 12]
```diff
@@ -641,6 +643,7 @@ func (s *Server) handleToken(w http.ResponseWriter, r *http.Request) {
 
 // handle an access token request https://tools.ietf.org/html/rfc6749#section-4.1.3
 func (s *Server) handleAuthCode(w http.ResponseWriter, r *http.Request, client storage.Client) {
+
```
Comment (JoelSpeed) [inline server/handlers.go]: Nit: Random new line

File: server/handlers.go [hunk 13]
```diff
@@ -660,7 +663,28 @@ func (s *Server) handleAuthCode(w http.ResponseWriter, r *http.Request, client s
...
+		Groups:        authCode.Claims.Groups,
+		ConnectorData: string(authCode.ConnectorData),
+		ConnectorId:   authCode.ConnectorID,
+		Expiry:        authCode.Expiry.Unix(),
+	}
+
+	// The accessToken is build from claims
+	accessToken, err := internal.Marshal(token)
+
+	// Encrypt access token to hide sensible data
```
Comment (JoelSpeed) [inline server/handlers.go]: Do we mean `sensible` here?
```suggestion
	// Encrypt access token to hide sensitive data
```

File: server/handlers.go [hunk 14]
```diff
@@ -974,6 +998,7 @@ func (s *Server) writeAccessToken(w http.ResponseWriter, idToken, accessToken, r
 	// TODO(ericchiang): figure out an access token story and support the user info
 	// endpoint. For now use a random value so no one depends on the access_token
 	// holding a specific structure.
+
```
Comment (JoelSpeed) [inline server/handlers.go]: Nit: Random new line?

File: server/handlers.go [hunk 15]
```diff
@@ -998,6 +1023,84 @@ func (s *Server) writeAccessToken(w http.ResponseWriter, idToken, accessToken, r
...
+	accessToken := new(internal.AccessToken)
+	if err := internal.Unmarshal(decryptedToken, accessToken); err != nil {
+		s.logger.Errorf("Failed to unmarshal access token: %v", err)
+	}
+
+	// Find the connector used in this auth flow
+	conn, err := s.getConnector(accessToken.ConnectorId)
+
+	if err != nil {
```
Comment (JoelSpeed) [inline server/handlers.go]: Nit: as above I would expect no new line between `conn, err := ` and `if err != nil`

File: server/handlers.go (hunk 5 above)
Comment (JoelSpeed) [reply server/handlers.go]: I was thinking exactly the same while reading through this, mutating `user` inside `GetUserInfo` is bound to lead to a bug somewhere down the line 😂

File: connector/oidc/oidc.go [hunk 16]
```diff
@@ -17,6 +18,11 @@ import (
 	"github.com/dexidp/dex/connector"
 )
 
+type connectorData struct {
+	// Store token, even if it eventually expires
```
Comment (JoelSpeed) [inline connector/oidc/oidc.go]: What happens currently when the token is expired? Do we return nothing? Do we return stale data?

File: connector/connector.go (hunk 2 above)
Comment (JoelSpeed) [reply connector/connector.go]: That's my understanding of this and I think that's probably the right thing to do, I was reading through [the standard claims](https://connect2id.com/products/server/docs/api/userinfo#claims) for the userinfo endpoint and there's a whole lot of info in there that Dex wouldn't normally have but upstream providers may provide somehow

File: server/handlers.go [hunk 17]
```diff
@@ -998,6 +1023,84 @@ func (s *Server) writeAccessToken(w http.ResponseWriter, idToken, accessToken, r
...
+	}
+
+	decryptedToken, err := s.Decrypt(token)
+	if err != nil {
+		s.logger.Errorf("Internal server error: %v", err)
+		s.renderError(w, http.StatusInternalServerError, "Internal server error.")
+		return
+	}
+
+	accessToken := new(internal.AccessToken)
```
Comment (JoelSpeed) [inline server/handlers.go]: I rarely see `new` used in Go, any reason not to just 
```suggestion
	accessToken := &internal.AccessToken{}
```

As we do with most other bare initialisations of structs?

File: server/security.go [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+	}
+
+	if keys.SigningKeyPub == nil {
+		return "", fmt.Errorf("no public keys found")
+	}
+
+	// Instantiate an encrypter using RSA-OAEP with AES128-GCM. An error would
+	// indicate that the selected algorithm(s) are not currently supported.
+	encrypter, err := jose.NewEncrypter(jose.A128GCM, jose.Recipient{Algorithm: jose.RSA_OAEP, Key: keys.SigningKeyPub}, nil)
+	// encrypter, err := jose.NewEncrypter(jose.RSA_OAEP, jose.A128GCM, &keys.SigningKeyPub)
```
Comment (JoelSpeed) [inline server/security.go]: What's this comment for? Debug leftover?

File: server/security.go (hunk 9 above)
Comment (JoelSpeed) [reply server/security.go]: Seems fairly sensible to me

**Comment by kalinon** [comment]
closing in deference to #1463

**Pull request closed without merging**